scripts/src/admin/wrapper.html

```diff
@@ -1,492 +1,510 @@
+		refreshPreview(overlayJsonEditor.get());
+	}
+
+	function refreshPreview(json) {
+		let jsonData = stringifyJson(json);
+		if (!jsonData) {
+			return;
+		}
+		previewElemnt.src = "${ webview }?cacheBuster=${ now }&jsonData=" + encodeURIComponent(jsonData);
+	}
+
+	function openEditor(jsonArea, appId) {
+		if (jsonArea.querySelector('.jsoneditor')) {
+			return false;
+		}
+		jsonArea.dataset.appId = appId;
+		jsonArea.querySelector('pre').hidden = true;
+
+		const editor = new JSONEditor(
+			jsonArea,
+			{
+				mode: 'form',
+				modes: ['view', 'form', 'tree', 'code']
+			},
+			JSON.parse(unescape(jsonArea.querySelector('pre').textContent))
+		);
+		editor.expandAll();
+		toast('Ctrl+Enter updates the config, Esc closes the editor');
+
+		jsonArea.querySelector('.jsoneditor').tabindex = 1;
+		jsonArea.querySelector('.jsoneditor').focus();
+		setTimeout(() => {
+			jsonArea.querySelector('.jsoneditor').addEventListener('keyup', e => {
+				attachEditorKeyboardEvents(editor, e)
+			})
+		}, 0)
+
+		const submitBtn = document.createElement('button');
+		submitBtn.textContent = 'Save ✓';
+		submitBtn.classList.add('editor-btn');
+		jsonArea.append(submitBtn);
+		submitBtn.addEventListener('click', e => {
+			e.stopPropagation();
+			confirmCbIfProduction(updateConfig.bind(this, editor));
+		});
+
+		const closeBtn = document.createElement('button')
+		closeBtn.textContent = '✖'
+		closeBtn.classList.add('editor-btn')
+		closeBtn.addEventListener('click', e => {
+			e.stopPropagation();
+			closeEditor(editor);
+		})
+		jsonArea.append(closeBtn)
+	}
+
+	async function updateConfig(editor) {
+		let json;
+		const appId = editor.frame.parentNode.dataset.appId;
+		try {
+			json = editor.get();
+			const isLimitsNumbers = (obj, limitName) => typeof obj.limits[limitName] === 'number';
+			if (!json.limits || !Object.keys(json.limits).every(isLimitsNumbers.bind(null, json))) {
+				toast('Invalid app limits format')
+				return false;
+			}
+
+			const response = await fetch(`/applications/${appId}/config`, {
+				method: 'PUT',
+				body: JSON.stringify(json),
+				headers: new Headers({'Content-Type': 'application/json'})
+			})
+			if (response.ok) {
+				closeEditor(editor)
+			} else {
+				toast('Config is incorrect (rejected by server)')
+			}
+		} catch (e) {
+			toast('smth is wrong');
+		}
+	}
+
+	function closeEditor(editor) {
+		const jsonArea = editor.frame.parentNode;
+		const configJson = JSON.stringify(editor.get(), null, 2);
+		editor.destroy();
+		Array.from(jsonArea.querySelectorAll('.editor-btn')).forEach(btn => {
+			btn.remove();
+		})
+		jsonArea.querySelector('pre').textContent = configJson;
+		jsonArea.querySelector('pre').removeAttribute('hidden')
+	}
+
+	function attachEditorKeyboardEvents(editor, e) {
+		switch (e.keyCode) {
+			case 13:
+				if (e.ctrlKey) {
+					updateConfig(editor)
+				}
+				break;
+			case 27:
+				closeEditor(editor);
+				break;
+		}
+	}
+
+	function confirmCbIfProduction(cb) {
```

Comment:
This is new

Reply:
ok. Didn't know about confirm. Does it work on Netscape?